File: app/src/main/java/com/mono/music/Recommend.kt
package com.mono.music

/** A titled row of suggestions on the For you tab. */
data class Section(val title: String, val subtitle: String, val songs: List<Song>)

/**
 * Builds the For you tab out of what is actually known about this library:
 * the tracks, the likes, and the play history. There is no service to ask, so a
 * suggestion here means "you have this, and here is why it is worth playing" —
 * what you keep going back to, what you have not heard in a while, and more by
 * the artists you liked.
 */
object Recommend {

    private const val ROW = 12

    fun sections(
        songs: List<Song>,
        favorites: Set<String>,
        playCounts: Map<String, Int>,
        lastPlayed: Map<String, Long>,
        day: Long = System.currentTimeMillis() / 86_400_000L,
    ): List<Section> {
        if (songs.isEmpty()) return emptyList()
        val out = mutableListOf<Section>()

        val recent = songs
            .filter { lastPlayed[it.docId] != null }
            .sortedByDescending { lastPlayed[it.docId] ?: 0L }
            .take(ROW)
        if (recent.isNotEmpty()) {
            out.add(Section("Recently played", "Pick up where you left off", recent))
        }

        val mostPlayed = songs
            .filter { (playCounts[it.docId] ?: 0) > 1 }
            .sortedByDescending { playCounts[it.docId] ?: 0 }
            .take(ROW)
        if (mostPlayed.isNotEmpty()) {
            out.add(Section("On repeat", "Your most played", mostPlayed))
        }

        val liked = songs.filter { it.docId in favorites }
        if (liked.isNotEmpty()) {
            out.add(Section("Liked songs", "${liked.size} tracks", liked))

            // More by the artists behind the likes — the tracks themselves
            // excluded, since those are one row up.
            val likedArtists = liked.map { it.artist.lowercase() }.toSet()
            val moreLikeThis = songs
                .filter { it.artist.lowercase() in likedArtists && it.docId !in favorites }
                .sortedByDescending { playCounts[it.docId] ?: 0 }
                .take(ROW)
            if (moreLikeThis.isNotEmpty()) {
                val names = liked.map { it.artist }.distinct()
                val because = if (names.size == 1) {
                    "Because you like ${names.first()}"
                } else {
                    "Based on what you liked"
                }
                out.add(Section("More like this", because, moreLikeThis))
            }
        }

        val added = songs.filter { it.addedAt > 0 }.sortedByDescending { it.addedAt }.take(ROW)
        if (added.isNotEmpty()) {
            out.add(Section("Recently added", "Newest in your folder", added))
        }

        // A rotating handful of tracks that have never been played. The window
        // moves with the day so the row is not the same every time the tab is
        // opened, without needing anything stored.
        val unplayed = songs.filter { playCounts[it.docId] == null }.sortedBy { it.title.lowercase() }
        if (unplayed.size > 2) {
            val start = ((day % unplayed.size) + unplayed.size) % unplayed.size
            val window = (0 until minOf(ROW, unplayed.size)).map {
                unplayed[((start + it) % unplayed.size).toInt()]
            }
            out.add(Section("Never played", "Buried in your library", window))
        }

        return out
    }
}

File: web/lib/recommend.js
/*
  The For you tab, ported from the Android app's Recommend.kt so both stay in
  step. There is no music service to ask, so a suggestion here means "you own
  this, and here is why it is worth playing": what you keep going back to, what
  you have not heard in a while, and more by the artists you liked.
*/

const ROW = 12;

export function sections({
  songs,
  likes,
  playCounts,
  lastPlayed,
  day = Math.floor(Date.now() / 86_400_000),
}) {
  if (!songs.length) return [];
  const out = [];
  const countOf = (song) => playCounts[song.path] || 0;

  const recent = songs
    .filter((song) => lastPlayed[song.path])
    .sort((a, b) => (lastPlayed[b.path] || 0) - (lastPlayed[a.path] || 0))
    .slice(0, ROW);
  if (recent.length) {
    out.push({ title: "Recently played", subtitle: "Pick up where you left off", songs: recent });
  }

  const mostPlayed = songs
    .filter((song) => countOf(song) > 1)
    .sort((a, b) => countOf(b) - countOf(a))
    .slice(0, ROW);
  if (mostPlayed.length) {
    out.push({ title: "On repeat", subtitle: "Your most played", songs: mostPlayed });
  }

  const liked = songs.filter((song) => likes.has(song.path));
  if (liked.length) {
    out.push({ title: "Liked songs", subtitle: `${liked.length} tracks`, songs: liked });

    // More by the artists behind the likes — the liked tracks themselves
    // excluded, since those are one row up.
    const likedArtists = new Set(liked.map((song) => song.artist.toLowerCase()));
    const moreLikeThis = songs
      .filter((song) => likedArtists.has(song.artist.toLowerCase()) && !likes.has(song.path))
      .sort((a, b) => countOf(b) - countOf(a))
      .slice(0, ROW);
    if (moreLikeThis.length) {
      const names = [...new Set(liked.map((song) => song.artist))];
      out.push({
        title: "More like this",
        subtitle: names.length === 1 ? `Because you like ${names[0]}` : "Based on what you liked",
        songs: moreLikeThis,
      });
    }
  }

  const added = songs
    .filter((song) => song.addedAt > 0)
    .sort((a, b) => b.addedAt - a.addedAt)
    .slice(0, ROW);
  if (added.length) {
    out.push({ title: "Recently added", subtitle: "Newest in your folder", songs: added });
  }

  // A rotating handful that have never been played. The window moves with the
  // day so the row is not identical every time the tab is opened, without
  // needing anything stored.
  const unplayed = songs
    .filter((song) => !playCounts[song.path])
    .sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()));
  if (unplayed.length > 2) {
    const start = ((day % unplayed.length) + unplayed.length) % unplayed.length;
    const window = [];
    for (let i = 0; i < Math.min(ROW, unplayed.length); i++) {
      window.push(unplayed[(start + i) % unplayed.length]);
    }
    out.push({ title: "Never played", subtitle: "Buried in your library", songs: window });
  }

  return out;
}
